app/drivers/kscan/kscan_gpio_matrix.c

```diff
@@ -1,48 +1,155 @@
+#define INST_COLS_LEN(n) DT_INST_PROP_LEN(n, col_gpios)
+#define INST_MATRIX_LEN(n) (INST_ROWS_LEN(n) * INST_COLS_LEN(n))
+#define INST_INPUTS_LEN(n) COND_DIODE_DIR(n, (INST_COLS_LEN(n)), (INST_ROWS_LEN(n)))
+
+#define USE_POLLING IS_ENABLED(CONFIG_ZMK_KSCAN_MATRIX_POLLING)
+#define USE_INTERRUPTS (!USE_POLLING)
+
+#define COND_INTERRUPTS(code) COND_CODE_1(CONFIG_ZMK_KSCAN_MATRIX_POLLING, (), code)
+#define COND_POLL_OR_INTERRUPTS(pollcode, intcode)                                                 \
+    COND_CODE_1(CONFIG_ZMK_KSCAN_MATRIX_POLLING, pollcode, intcode)
+
+// TODO (Zephr 2.6): replace the following
+// kscan_gpio_dt_spec -> gpio_dt_spec
+// KSCAN_GPIO_DT_SPEC_GET_BY_IDX -> GPIO_DT_SPEC_GET_BY_IDX
+// gpio_pin_get -> gpio_pin_get_dt
+// gpio_pin_set -> gpio_pin_set_dt
+// gpio_pin_interrupt_configure -> gpio_pin_interrupt_configure_dt
+struct kscan_gpio_dt_spec {
+    const struct device *port;
     gpio_pin_t pin;
-    gpio_flags_t flags;
+    gpio_dt_flags_t dt_flags;
 };
 
-#define _KSCAN_GPIO_ITEM_CFG_INIT(n, prop, idx)                                                    \
+#define KSCAN_GPIO_DT_SPEC_GET_BY_IDX(node_id, prop, idx)                                          \
     {                                                                                              \
-        .label = DT_INST_GPIO_LABEL_BY_IDX(n, prop, idx),                                          \
-        .pin = DT_INST_GPIO_PIN_BY_IDX(n, prop, idx),                                              \
-        .flags = DT_INST_GPIO_FLAGS_BY_IDX(n, prop, idx),                                          \
-    },
+        .port = DEVICE_DT_GET(DT_GPIO_CTLR_BY_IDX(node_id, prop, idx)),                            \
+        .pin = DT_GPIO_PIN_BY_IDX(node_id, prop, idx),                                             \
+        .dt_flags = DT_GPIO_FLAGS_BY_IDX(node_id, prop, idx),                                      \
+    }
+
+#define KSCAN_GPIO_ROW_CFG_INIT(idx, inst_idx)                                                     \
+    KSCAN_GPIO_DT_SPEC_GET_BY_IDX(DT_DRV_INST(inst_idx), row_gpios, idx),
+#define KSCAN_GPIO_COL_CFG_INIT(idx, inst_idx)                                                     \
+    KSCAN_GPIO_DT_SPEC_GET_BY_IDX(DT_DRV_INST(inst_idx), col_gpios, idx),
+
+enum kscan_diode_direction {
+    KSCAN_ROW2COL,
+    KSCAN_COL2ROW,
+};
+
+struct kscan_matrix_irq_callback {
+    const struct device *dev;
+    struct gpio_callback callback;
+    struct k_delayed_work *work;
+};
+
+struct kscan_matrix_data {
+    const struct device *dev;
+    kscan_callback_t callback;
+    struct k_delayed_work work;
+#if USE_POLLING
+    struct k_timer poll_timer;
+#else
+    /** Array of length config->inputs.len */
+    struct kscan_matrix_irq_callback *irqs;
+#endif
+    /** Current state of the matrix as a flattened 2D array. */
+    bool *current_state;
+    /** Temporary buffer for reading in the next matrix state. */
```

Comment:
Can we note the 2D/length details here as well please?

Reply:
Done.